Scijava/scijava ops/hints

gselzer opened the pull request:
This PR provides a two-way mechanism to deliver matching preferences (called hints) to the matcher. Hints can be declared by:
* Ops, using the [`OpHints`](https://github.com/scijava/incubator/blob/4efe3f8025a0bee0dea59d016f3e65376ed69120/scijava/scijava-ops/src/main/java/org/scijava/ops/hints/OpHints.java#L16) annotation. This data is stored within the `OpInfo`.
* Users, using the [`Hints`](https://github.com/scijava/incubator/blob/scijava/scijava-ops/hints/scijava/scijava-ops/src/main/java/org/scijava/ops/hints/Hints.java) interface. A `Hints` object can be created by the user or can be obtained through an `OpEnvironment`

These hints can then be used to indicate their matching preferences. For example, Ops can use the hint [`Simplification.FORBIDDEN`](https://github.com/scijava/incubator/blob/4efe3f8025a0bee0dea59d016f3e65376ed69120/scijava/scijava-ops/src/main/java/org/scijava/ops/hints/BaseOpHints.java#L16) to declare that it should not be simplified. Conversely, a user can pass the same hint to the matcher via a `Hints` object to specify that they do not want to match simplified Ops.

This PR also moves the simplification matching logic (`SimplifiedOpRef`/`SimplifiedOpInfo` construction, etc.) to `DefaultOpMatcher` (solves part of 

Closes scijava/scijava-ops#43.

TODO:
* [ ] Determine whether any other Hints should be supported out-of-the-box (@ctrueden @hinerm would you like to see any other hints available?)
* [x] Rebase over latest `master`

gselzer committed:
First cut at a Hints data structure

gselzer committed:
Update OpInfo infrastructure to hold onto hints

gselzer committed:
Refactor DefaultOpEnvironment to use hints

gselzer committed:
Add API for passing Hints to OpEnvironment

gselzer committed:
Copy Hints in setHints

We do not want to allow users to change the hints Object that the
Environment's hints points to by changing their own.

gselzer committed:
Extract similar logic to AbstractHints

This also allows us to write an equals() method that is identical for
all Hints implementations (essentially just comparing the Hints maps).

gselzer committed:
Big Commit: Add API, spread hints through matcher

gselzer committed:
Expose hints package

This allows users to access Hints. This commit also adds a hints.impl
package to prevent the exposure of the implementations

gselzer committed:
Export adaptation/simplification hints

These hints are so fundamental that users will want to use them

gselzer committed:
Test simplification hints

gselzer committed:
Remove Unsimplifiable, use hints instead

gselzer committed:
Replace copyOp warning with javadoc

gselzer committed:
Move simplification matching logic

This logic belongs better within the Matcher

gselzer committed:
Cast original Op output to the correct type

This instantiation of originalOut could cause CannotCompileExceptions
without the cast there to assure Javassist's compiler that the op return
is of the right type.

This commit also extracts the Op calling additions to the StringBuilder
to a separate method

gselzer commented:
Assuming that we get `therapi-runtime-javadoc` working via scijava/scijava-ops#50, it might be worth considering a `Hint` to disable the javadoc scraping. Unsure if it is computationally expensive enough to warrant a hint, but definitely worth keeping an eye on.

hinerm commented on scijava/scijava-ops/src/main/java/org/scijava/ops/OpInfo.java:
```diff
@@ -29,9 +32,14 @@
-	boolean isSimplifiable();
+
+	/** Gets the hints declared in the {@link OpHints} annotation */
+	List<String> declaredHints();
```
Do we want to lose the connection to `OpHints` here? Is there any advantage to unwrapping it to a `List<String>`?

hinerm replied:
OK, @gselzer: create a new `Hints` implementation that's read-only and backed by `ImmutableList`.

gselzer replied:
Done!

hinerm replied:
@gselzer I think we may want to return the actual `ImmutableHints` here, to concretely signal that the object is unmodifiable.  It's OK to have specific return types.

gselzer replied:
@hinerm this is true, my only thought was that maybe there are times where it is too restrictive to return an `ImmutableHints`. But I can't think of a good use case, so I agree that we can make it `Immutable`

gselzer replied:
@hinerm, ah, I forgot that the other reason for returning `Hints`: to return `ImmutableHints`, we'd have to `export` it. Is this something we want to do? Furthermore, since `ImmutableHints` extends `AbstractHints`, we'd have to export that as well, which I like less. I'd prefer to only expose the interfaces, what do you think of this?

hinerm replied:
@gselzer that's fair

hinerm commented on scijava/scijava-ops/src/main/java/org/scijava/ops/matcher/OpMethodInfo.java:
```diff
@@ -78,7 +78,7 @@
 	private final ValidityException validityException;
 
-	private final boolean simplifiable;
+	private final List<String> hints;
```
Same question - why a `List<String>` vs `OpHints`?

hinerm commented on scijava/scijava-ops/src/main/java/org/scijava/ops/OpEnvironment.java:
```diff
@@ -139,4 +147,16 @@ default OpBuilder op(final String opName) {
+	 * 
+	 * @param hints
+	 */
+	void setHints(Hints hints);
```
TODO: add a disclaimer that to guarantee the `Hints` that are used during an op call, you should include the `Hints` in the call instead of at this environment level.

gselzer replied:
Done!

hinerm commented on scijava/scijava-ops/src/test/java/org/scijava/ops/hints/AdaptationHintTest.java:
```diff
@@ -0,0 +1,72 @@
+		hints.setHint(Adaptation.ALLOWED);
+		ops.env().setHints(hints);
+		@SuppressWarnings("unused")
+		Computers.Arity1<Double[], Double[]> adaptable = ops.op(
```
@gselzer Could you add a `Hints` signature to the final methods of `OpBuilder` so that you can build an `Op` call with a particular set of `Hints`? (instead of requiring it to be set at the environment level)

gselzer replied:
Done!

gselzer committed:
Reform OpInfo hints into a Hints Object

This better preserves the idea has a set of Hints, compared to the
List<String> it was before. Since the user should not be able to change
the Hints a OpInfo declares, we create a new Hints type, ImmutableHints,
which all OpInfos should use.

gselzer committed:
Add javadoc specifying setHint unsafety

gselzer committed:
Add hints option to OpBuilders

gselzer committed:
Test passing Hints via OpBuilder call

gselzer committed:
Fix Default/AbstractHints javadoc
